### PR TITLE
Ingest stories, quickly return via API responses

### PR DESCRIPTION
**Pull request by kookster**
- [x] Add in actual db columns and store the story data and any overrides
- [x] Q: Do we copy the audio for feeder to use, or keep using the CMS urls?
- [x] Check that story update/delete notifications are more recent than last used to update the episode
- [x] Make sure the feeder API calls used by dovetail are cached appropriately

### CODE REVIEW

**Comment by kookster** [comment]
Hey @gcampo88 wanna take a look at a railsy PR?

**Comment by kookster** [comment]
@cavis sort of a huge PR, sorry 😏

**Comment by cavis** [comment]
Less than 30K lines of changes?  I'm disappointed.

File: app/models/episode_story_handler.rb [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      episode.lock!
+      update_from_story(story)
+      episode.save!
+    end
+    episode
+  end
+
+  def update_from_story(story)
+    self.story = story
+    self.episode.prx_uri = story.links['self'].href
```
Comment (cavis) [inline app/models/episode_story_handler.rb]: Any worries about relying on these CMS-generated urls?  (What if the api version changes?)

File: app/models/episode_story_handler.rb [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+  def update_attributes
+    sa = story.attributes
+    self.episode.title = sa[:title]
+    self.episode.subtitle = Sanitize.fragment(sa[:short_description] || '').strip
+    self.episode.description = Sanitize.fragment(sa[:description] || '').strip
+    self.episode.summary = sa[:description]
+    self.episode.content = sa[:description]
+    self.episode.categories = sa[:tags]
+    self.episode.published_at = Time.parse(sa[:published_at]) if sa[:published_at]
```
Comment (cavis) [inline app/models/episode_story_handler.rb]: Aren't announced attributes camelCase?

File: app/models/episode_story_handler.rb [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    self.episode.updated = Time.parse(sa[:updated_at]) if sa[:updated_at]
+  end
+
+  def update_audio
+    audio = story.objects["prx:audio"].objects['prx:items'] rescue []
+    audio.each do |a|
+      next unless a.status == 'complete'
```
Comment (cavis) [inline app/models/episode_story_handler.rb]: I'm guessing this is the magic that prevents publishing an episode before the media completes processing?

File: app/models/episode_story_handler.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    self.episode.description = Sanitize.fragment(sa[:description] || '').strip
+    self.episode.summary = sa[:description]
+    self.episode.content = sa[:description]
+    self.episode.categories = sa[:tags]
+    self.episode.published_at = Time.parse(sa[:published_at]) if sa[:published_at]
+    self.episode.updated = Time.parse(sa[:updated_at]) if sa[:updated_at]
+  end
+
+  def update_audio
+    audio = story.objects["prx:audio"].objects['prx:items'] rescue []
```
Comment (cavis) [inline app/models/episode_story_handler.rb]: What would Hound say?

File: db/migrate/20161014123500_add_episode_attributes.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    add_column :episodes, :description, :text
+    add_column :episodes, :categories, :text
+    add_column :episodes, :block, :boolean
+    add_column :episodes, :is_closed_captioned, :boolean
+    add_column :episodes, :position, :integer
+    add_column :episodes, :feedburner_orig_link, :string
+    add_column :episodes, :feedburner_orig_enclosure_link, :string
+    add_column :episodes, :is_perma_link, :boolean
+
+    Episode.where(prx_uri: nil, title: nil).each { |e| EpisodeEntryHandler.new(e).update_from_overrides }
```
Comment (cavis) [inline db/migrate/20161014123500_add_episode_attributes.rb]: So `overrides` is sticking around, even with all these normalized columns?

File: test/fixtures/prx_story_all.json [hunk 6]
```diff
@@ -0,0 +1,609 @@
...
+          "href": "/api/v1/accounts/208391/stories{?page,per,zoom,filters,sorts}",
+          "templated": true
+        },
+        "self": {
+          "href": "/api/v1/accounts/208391",
+          "profile": "http://meta.prx.org/model/account/individual/min"
+        }
+      },
+      "id": 208391,
+      "name": "Ryan Cavis",
```
Comment (cavis) [inline test/fixtures/prx_story_all.json]: Fixture'd.

File: app/models/episode_story_handler.rb (hunk 3 above)
Comment (cavis) [inline app/models/episode_story_handler.rb]: What exactly prevents an episode from publishing if some-but-not-all of the audio is complete?

Also - I'm not finding any references to the story-image.  Is that included in the feed yet?

File: app/jobs/story_update_job.rb [hunk 7]
```diff
@@ -42,12 +42,17 @@ def load_resources(data)
   end
 
   def update_episode
-    episode.restore if episode.deleted?
-    episode.update_from_story!(story)
+    story_updated = Time.parse(story.attributes[:updated_at]) if story.attributes[:updated_at]
+    if episode.updated && story_last_updated && story_updated < episode.updated
```
Comment (kookster) [inline app/jobs/story_update_job.rb]: yes, sqs is terrible about order of delivery

File: app/models/episode_story_handler.rb [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+  attr_accessor :episode, :story
+
+  def initialize(episode)
+    self.episode = episode
+  end
+
+  def self.create_from_story!(story)
+    series_uri = story.links['series'].href
+    podcast = Podcast.find_by!(prx_uri: series_uri)
```
Comment (kookster) [inline app/models/episode_story_handler.rb]: currently we would have to set it up ahead of time, but soon I want to have a better way.

File: app/models/episode_story_handler.rb (hunk 1 above)
Comment (kookster) [inline app/models/episode_story_handler.rb]: do you have a better suggestion what we should store to correlate with the CMS stories?
Didn't seem like it would be very hypermedia to use the story id, but we do have that

File: app/models/episode_story_handler.rb (hunk 4 above)
Comment (kookster) [inline app/models/episode_story_handler.rb]: hound loves mismatched quotes, c'mon, it's creative! (I'll fix it)

File: app/models/episode_story_handler.rb (hunk 3 above)
Comment (kookster) [inline app/models/episode_story_handler.rb]: story images, real any rss item images, not in the feed yet, will be another PR for that.

File: db/migrate/20161014123500_add_episode_attributes.rb (hunk 5 above)
Comment (kookster) [inline db/migrate/20161014123500_add_episode_attributes.rb]: I need it around for this PR/deploy, but I am considering getting rid of it after that, but I wanted to make sure I could upgrade to this with it in place, and revert back to using it if necessary

File: test/fixtures/prx_story_all.json (hunk 6 above)
Comment (kookster) [inline test/fixtures/prx_story_all.json]: (in)famous!

**Comment by kookster** [comment]
re: feeder_default queue, that is not used for this - announce creates its own sns and sqs queues, and subscribes to them based on the announce.yml config, so story CRUD messages come over specific announce queues

**Comment by kookster** [comment]
@cavis I believe I fixed all the things

File: app/models/episode_story_handler.rb (hunk 1 above)
Comment (cavis) [inline app/models/episode_story_handler.rb]: Nah, no need to pre-engineer a solution - just want to make sure we're all aware of it.

**Comment by cavis** [comment]
👍 all the things
